Use repeatable read transaction isolation level

tolikzinovyev opened the pull request:
Use database connections instead of transactions to support multiple databases behind a loadbalancer (#351). Also fixes #391.

brianolson commented:
This might be fine but I'm not sure it solves the previous problem. A connection doesn't have higher default transaction serializeability guarantee, only an explicit transaction has that (in the Go database/sql API, apparently). So I think this leaves us open to the previously seen skew of different queries within an Indexer-API potentially having inconsistent results. But, maybe we're giving up because that's inevitable because Postgres read-only-replicas don't allow transactions anyway?

tolikzinovyev commented:
All we need is receiving consistent results for individual queries and receiving database state version in the increasing order. For example, if I now that round x is accounted, then I should be able to get transactions for round x. Doesn't a read-only replica process updates atomically and in order? I would expect so.

winder commented:
We have other types of queries with race conditions. For example:
1. Query the accounting round, receive round N
2. Accounting completes, updates the account and accounting round.
3. Query account data, receive data for round N+1.

I think you're right in that we could find a way to consistently use connections to fetch transactions, but then we would have to use the max round filter for all queries (or remove transactions at round N+1).

What is the motivation to using connections over transactions?

tolikzinovyev commented:
@winder you are right. I kind of took for granted that the code before we added transactions worked correctly with a single database. You provided a counter-example. I changed connections to repeatable read transactions, they provide "snapshot-like" read consistency, so that should be sufficient.

tolikzinovyev commented:
And I don't know if transactions impact performance here really. When you query blockchain transactions, the database needs to keep a copy of them, but they never change, so there is probably no cost.

winder reviewed (approved):
The repeatable read change looks good
